### PR TITLE
 [SPARK-31692][SQL][2.4] Pass hadoop confs  specifed via Spark confs to URLStreamHandlerfactory

### PR DESCRIPTION
**Pull request by karuppayya**
### What changes were proposed in this pull request?
Pass hadoop confs  specifed via Spark confs to URLStreamHandlerfactory

### Why are the changes needed?

**BEFORE**
```
➜  spark git:(SPARK-31692) ✗ ./bin/spark-shell --conf spark.hadoop.fs.file.impl=org.apache.hadoop.fs.RawLocalFileSystem

scala> spark.sharedState
res0: org.apache.spark.sql.internal.SharedState = org.apache.spark.sql.internal.SharedState@5793cd84

scala> new java.net.URL("file:///tmp/1.txt").openConnection.getInputStream
res1: java.io.InputStream = org.apache.hadoop.fs.ChecksumFileSystem$FSDataBoundedInputStream@22846025

scala> import org.apache.hadoop.fs._
import org.apache.hadoop.fs._

scala>  FileSystem.get(new Path("file:///tmp/1.txt").toUri, spark.sparkContext.hadoopConfiguration)
res2: org.apache.hadoop.fs.FileSystem = org.apache.hadoop.fs.LocalFileSystem@5a930c03
```

**AFTER**
```
➜  spark git:(SPARK-31692) ✗ ./bin/spark-shell --conf spark.hadoop.fs.file.impl=org.apache.hadoop.fs.RawLocalFileSystem

scala> spark.sharedState
res0: org.apache.spark.sql.internal.SharedState = org.apache.spark.sql.internal.SharedState@5c24a636

scala> new java.net.URL("file:///tmp/1.txt").openConnection.getInputStream
res1: java.io.InputStream = org.apache.hadoop.fs.FSDataInputStream@2ba8f528

scala> import org.apache.hadoop.fs._
import org.apache.hadoop.fs._

scala>  FileSystem.get(new Path("file:///tmp/1.txt").toUri, spark.sparkContext.hadoopConfiguration)
res2: org.apache.hadoop.fs.FileSystem = LocalFS

scala>  FileSystem.get(new Path("file:///tmp/1.txt").toUri, spark.sparkContext.hadoopConfiguration).getClass
res3: Class[_ <: org.apache.hadoop.fs.FileSystem] = class org.apache.hadoop.fs.RawLocalFileSystem
```
The type of FileSystem object created(you can check the last statement in the above snippets) in the above two cases are different, which should not have been the case

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

Pass the jenkins with newly added test cases.

### CODE REVIEW

**Comment by karuppayya** [comment]
Backport commit for SPARK-31692 @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Thank you, @karuppayya !

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala [hunk 1]
```diff
@@ -157,7 +157,13 @@ private[sql] class SharedState(val sparkContext: SparkContext) extends Logging {
 
 object SharedState extends Logging {
   try {
-    URL.setURLStreamHandlerFactory(new FsUrlStreamHandlerFactory())
+    SparkSession.getActiveSession match {
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:160]: Ur, is it the same with branch-3.0?

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:160]: Is this because we don't have a configuration patch in branch-2.4?

Comment (karuppayya) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:160]: Yes, we dont have the configuration patch

**Comment by dongjoon-hyun** [comment]
Hmm. In this case, I believe we had better have the conf patch because we can disable it when we have a regression on this PR.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by dongjoon-hyun** [comment]
Thank you for your patience, @karuppayya .
For this PR, I backported two commits additionally as a preparation.

**Comment by dongjoon-hyun** [comment]
@karuppayya . I cherry-picked your original commit from `branch-3.0` to `branch-2.4` and resolved the conflicts. Thanks!

**Pull request closed without merging**